controllers/csi.go

```diff
@@ -0,0 +1,689 @@
+	args := []string{
+		fmt.Sprintf("--csi-address=%s", defaultSocketPath),
+		"--v=5",
+		"--timeout=150s",
+		"--leader-election=true",
+		fmt.Sprintf("--leader-election-namespace=%s", s.namespace),
+		"--handle-volume-inuse-error=false",
+	}
+
+	resourceRequirements := corev1.ResourceRequirements{
+		Limits: corev1.ResourceList{
+			// Empty for now
+		},
+		Requests: corev1.ResourceList{
+			// Empty for now
+		},
+	}
+
+	volumeMounts := []corev1.VolumeMount{
+		{Name: "socket-dir", MountPath: defaultSockerDir},
+	}
+
+	env := []corev1.EnvVar{
+		// Empty for now
+	}
+
+	csiResizer := &corev1.Container{
+		Name:            "csi-resizer",
+		Image:           defaultResizerImage,
+		Args:            args,
+		Resources:       resourceRequirements,
+		VolumeMounts:    volumeMounts,
+		ImagePullPolicy: corev1.PullIfNotPresent,
+		Env:             env,
+	}
+
+	return csiResizer
+}
+
+func (s *sideCarContainer) getCsiAttacherContainer() *corev1.Container {
+	// csi attacher container
+	args := []string{
+		fmt.Sprintf("--csi-address=%s", defaultSocketPath),
+		"--v=5",
+		"--timeout=150s",
+		"--leader-election=true",
+		fmt.Sprintf("--leader-election-namespace=%s", s.namespace),
+	}
+
+	resourceRequirements := corev1.ResourceRequirements{
+		Limits: corev1.ResourceList{
+			// Empty for now
+		},
+		Requests: corev1.ResourceList{
+			// Empty for now
+		},
+	}
+
+	volumeMounts := []corev1.VolumeMount{
+		{Name: "socket-dir", MountPath: defaultSockerDir},
+	}
+
+	env := []corev1.EnvVar{
+		// Empty for now
+	}
+
+	csiAttacher := &corev1.Container{
+		Name:            "csi-attacher",
+		Image:           defaultAttacherImage,
+		Args:            args,
+		Resources:       resourceRequirements,
+		VolumeMounts:    volumeMounts,
+		ImagePullPolicy: corev1.PullIfNotPresent,
+		Env:             env,
+	}
+
+	return csiAttacher
+}
+
+func (s *sideCarContainer) getCsiSnapshotterContainer() *corev1.Container {
+	// csi snapshotter container
+	args := []string{
+		fmt.Sprintf("--csi-address=%s", defaultSocketPath),
+		"--v=5",
+		"--timeout=150s",
+		"--leader-election=true",
+		fmt.Sprintf("--leader-election-namespace=%s", s.namespace),
+		"--extra-create-metadata=true",
+	}
+
+	resourceRequirements := corev1.ResourceRequirements{
+		Limits: corev1.ResourceList{
+			// Empty for now
+		},
+		Requests: corev1.ResourceList{
+			// Empty for now
+		},
+	}
+
+	volumeMounts := []corev1.VolumeMount{
+		{Name: "socket-dir", MountPath: defaultSockerDir},
+	}
+
+	env := []corev1.EnvVar{
+		// Empty for now
+	}
+
+	csiSnapshotter := &corev1.Container{
+		Name:            "csi-snaphotter",
+		Image:           defaultSnapshotterImage,
+		Args:            args,
+		Resources:       resourceRequirements,
+		VolumeMounts:    volumeMounts,
+		ImagePullPolicy: corev1.PullIfNotPresent,
+		Env:             env,
+	}
+
+	return csiSnapshotter
+}
+
+func (s *sideCarContainer) getCephCsiContainer(pluginType, driverName string, controller bool) *corev1.Container {
```

Comment:
I think you could trim this down even further. Passing a `bool` into a function is often a sign that you could probably simplify things. I would get rid of `controller` and just append to the VolumeMounts outside this function.

Reply:
the controller is also used to set `fmt.Sprintf("--controllerserver=%t", controller)` argument if we want to take it out we need to set this argument in `4` places (rbddeployment,cephfsdeployment,rbddaemonset,cephfsdaemonset). For volumeMount again we need to do it at `2` places for daemonset. let me know WDYT

Reply:
Ah, okay... I think I could see a way to remove the boolean without increasing complexity too much, but your argument is valid for now.